src/ext/service.rs

```diff
@@ -99,65 +98,69 @@ impl ServiceExt for dc::Service {
                        sources: &'b sources::Sources)
                        -> Result<Sources<'b>> {
         // Get our `context`, if any.
-        let source_mount_dir = self.source_mount_dir()?;
-        let repo_subdir = self.repo_subdir()?;
-        let context = self.context()?.map(|ctx| (source_mount_dir, repo_subdir, ctx.clone()));
+        let container_path = self.source_mount_dir()?;
+        let source_subdirectory = self.source_subdirectory()?;
+
+        let context = self.context()?
+            .and_then(|ctx| {
+                // human_alias is called on every context when constructing Sources
+                let alias = &ctx.human_alias().expect("human_alias failed on a context that worked previously");
+                sources.find_by_alias(alias)
+            })
+            .and_then(|source| {
+                Some(SourceMount { container_path, source, source_subdirectory })
+            });
 
         // Get our library keys and mount points.
         let mut libs = vec![];
         for (label, mount_as) in &self.labels {
             let prefix = "io.fdy.cage.lib.";
             if label.starts_with(prefix) {
-                libs.push((mount_as.value()?.to_owned(),
-                           (&label[prefix.len()..]).to_owned()));
+                let lib_name = label[prefix.len()..].to_string();
+                let source: Result<&Source> = sources
+                    .find_by_lib_key(&lib_name)
+                    .ok_or_else(|| ErrorKind::UnknownLibKey(lib_name).into());
+
+                libs.push(SourceMount {
+                    container_path: mount_as.value()?.to_owned(),
+                    source: source?,
+                    source_subdirectory: None,
+                })
             }
         }
 
-        Ok(Sources {
-            sources: sources,
-            context: context,
-            libs: libs.into_iter(),
-        })
+        Ok(Sources { context, libs: libs.into_iter() })
     }
 }
 
 /// Iterator over all the `Source` trees which can be mounted into this
 /// `Service`.
 pub struct Sources<'a> {
-    /// All `Source` trees available for this repository.
-    sources: &'a sources::Sources,
     /// Do we need to iterate over our `context` field?
-    context: Option<(String, Option<String>, dc::Context)>,
+    context: Option<SourceMount<'a>>,
     /// Libraries
-    libs: vec::IntoIter<(String, String)>,
+    libs: vec::IntoIter<SourceMount<'a>>,
+}
+
+#[derive(Clone)]
+pub struct SourceMount<'a> {
+    pub container_path: String,
+    pub source: &'a Source,
+    pub source_subdirectory: Option<String>,
 }
 
 impl<'a> Iterator for Sources<'a> {
-    type Item = Result<(String, Option<String>, &'a Source)>;
+    type Item = SourceMount<'a>;
```

Comment:
More type unification, the `Item` is now the same as the fields of the struct we're iterating over.